utils/env_preferences/src/linux.rs

```diff
@@ -0,0 +1,112 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+pub mod linux_prefs {
+    use std::{borrow::Cow, collections::HashMap, ffi::CStr, ptr};
+
+    use libc::{setlocale, LC_ALL, LC_TIME};
+    use std::str::FromStr;
+
+    #[derive(Hash, Eq, PartialEq, Debug)]
+    pub enum LocaleCategory {
+        Character,
+        Number,
+        Time,
+        Collate,
+        Monetary,
+        Messages,
+        Paper,
+        Name,
+        Address,
+        Telephone,
+        Measurement,
+        Identification,
+        All,
+        Other(String),
+    }
+
+    impl FromStr for LocaleCategory {
+        type Err = ();
+
+        fn from_str(s: &str) -> Result<Self, Self::Err> {
+            match s {
+                "LC_CTYPE" => Ok(LocaleCategory::Character),
+                "LC_NUMERIC" => Ok(LocaleCategory::Number),
+                "LC_TIME" => Ok(LocaleCategory::Time),
+                "LC_COLLATE" => Ok(LocaleCategory::Collate),
+                "LC_MONETARY" => Ok(LocaleCategory::Monetary),
+                "LC_MESSAGES" => Ok(LocaleCategory::Messages),
+                "LC_PAPER" => Ok(LocaleCategory::Paper),
+                "LC_NAME" => Ok(LocaleCategory::Name),
+                "LC_ADDRESS" => Ok(LocaleCategory::Address),
+                "LC_TELEPHONE" => Ok(LocaleCategory::Telephone),
+                "LC_MEASUREMENT" => Ok(LocaleCategory::Measurement),
+                "LC_IDENTIFICATION" => Ok(LocaleCategory::Identification),
+                "LC_ALL" => Ok(LocaleCategory::All),
+                _ => Ok(LocaleCategory::Other(s.to_string())),
+            }
+        }
+    }
+
+    pub fn get_locales() -> HashMap<LocaleCategory, String> {
+        let mut locale_map = HashMap::new();
+
+        // SAFETY: Safety is ensured because we pass a `NULL` pointer and retrieve the locale there is
+        // no subsequent calls for `setlocale` which could change the locale of this particular thread
+        let locales_ptr = unsafe { setlocale(LC_ALL, ptr::null()) };
+
+        if locales_ptr.is_null() {
+            locale_map.insert(LocaleCategory::All, "C".to_string());
```

Comment:
low level - why are you adding a value here? If locale_ptr is null, locale_map should return null here.